### PR TITLE
gitea: new options and settings refactor

### PR DESCRIPTION
**Pull request by ThinkChaos**
###### Description of changes

A bunch of small improvements to the Gitea module.  
Highlights:
- make service group configurable
- add `customDir` option
  I use this to prevent the config which contains secrets (and the other secret files) from being stored in a persistent folder by using `customDir=/run/gitea/custom`.
- convert more simple settings to freeform
  Two commits have this goal. I split the second one since it requires changes to `pkgs.formats.ini` so I can easily revert it if needed.
  The main `formats` change is to support an optional global section. I added a test.
  I also needed to add support for append text to the result to support `extraConfig`. I don't think this is the best approach, and will add a review comment about this.

All changes are backwards compatible. `enableUnixSocket` could be removed instead of deprecated if back-compat is not worth the (minimal) extra complexity.  
Also I think we should drop the `extraConfig` as it adds complexity to the `formats` that we don't need here.

The `lfs` options could also be replaced by options under `settings`. I didn't do that since I think those Gitea settings aren't really clear. So, IMO, the NixOS option adds value. Let me know if you'd like me to also rename them.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/pkgs-lib/formats.nix
Comment (ThinkChaos) [inline pkgs/pkgs-lib/formats.nix:98]: I think having this as an argument to the generator isn't the best, but just a quick hack so I can get feedback.

My preferred solution would just be to remove `services.gitea.extraConfig` and not have to deal with this at all. For Gitea since the config format is plain ini, `extraConfig` has very little value since Nix can already represent everything that `extraConfig` can.

I do see the value in not breaking existing configs though, so if we want to keep the `extraConfig` option, I think this would be better implemented by adding a function to the format. Something like `generateText` which would do what `generate` does but without creating a file, just returning a string.  

Basically:
```nix
{
  generateText = value: ...;
  generate = name: value: pkgs.writeText name (generateText value);
}
```

Seeing as formats where defined with an RFC, I wanted to get feedback before doing that.
Adding `generateText` is far from as major as introducing formats, so I didn't think I'd create an RFC just yet. And this is my first real Nix(OS) contribution so that'd be a bit much :smile:

Comment (RaitoBezarius) [reply pkgs/pkgs-lib/formats.nix:98]: In order to remove `extraConfig`, we need to deprecate it at least for one release cycle. I don't recall if this has been done or not.

Under the assumption, we cannot, I feel like having this `extraConfig` is a bit much, but I would defer to @roberth @infinisil who may have better opinions on this.

Comment (ThinkChaos) [reply pkgs/pkgs-lib/formats.nix:98]: Yeah definitely understand we'd want to deprecate it first. So I think the `generateText` is the best realistic approach and would appreciate feedback on it.  
@infinisil is probably one the best people to do so, thanks for mentioning them!

Comment (roberth) [reply pkgs/pkgs-lib/formats.nix:98]: I'd prefer to keep the `generate` name, so as not to change the "type" of generators.

By keeping things "regular", it can be handled better by code.

I don't foresee any real problems in the most complicated use case I'm aware of, below, but it does illustrate that sticking to the standard is safer.

https://github.com/hercules-ci/flake.parts-website/blob/fa9f76d7168c0ad02abec0fcdbf2ce16d23681d3/render/render-module.nix#L24-L40

Comment (ThinkChaos) [reply pkgs/pkgs-lib/formats.nix]: Maybe my comment was unclear, but `generateText` would be in addition to `generate`, to keep the "type" backwards compatible.  
See https://github.com/NixOS/nixpkgs/pull/206179/commits/55399f49ca24dead0bde1a87e860735993c6d65f for an implementation of the idea, hopefully an diff is worth a 1000 words!

File: pkgs/pkgs-lib/formats.nix
Comment (SuperSandro2000) [inline pkgs/pkgs-lib/formats.nix:155]:
```suggestion
      in
        if globalSection then
          lib.generators.toINIWithGlobalSection genArgs {
            globalSection = filterSections false;
            sections = filterSections true;
          }
        else
          lib.generators.toINI genArgs transformedValue;
```

File: nixos/modules/services/misc/gitea.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/gitea.nix:652]:
```suggestion
      optional cfg.enableUnixSocket "services.gitea.enableUnixSocket is deprecated, please use services.gitea.settings.PROTOCOL = \"http+unix\";." ++
```

File: nixos/modules/services/misc/gitea.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/gitea.nix:548]:
```suggestion
            if [ ! -s '${lfsJwtSecret}' ]; then
                ${gitea}/bin/gitea generate secret LFS_JWT_SECRET > '${lfsJwtSecret}'
            fi
```

File: nixos/modules/services/misc/gitea.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/gitea.nix:493]: recursively chmodding and chowning all files especially under repository can slow down tmpfiles.d a lot depending on the size of your gitea installation.

File: nixos/modules/services/misc/gitea.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/gitea.nix:327]:
```suggestion
                defaultText = literalExpression "config.${opt.package}.data";
```

File: nixos/modules/services/misc/gitea.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/gitea.nix:319]: How does this handle committing the default port for http and https?

File: nixos/modules/services/misc/gitea.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/gitea.nix:274]:
```suggestion
              description = lib.mdDoc "Application title shown on the front page.";
```

File: nixos/modules/services/misc/gitea.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/gitea.nix:234]: Why not deprecate it then?

File: pkgs/pkgs-lib/formats.nix
Comment (roberth) [reply pkgs/pkgs-lib/formats.nix]: > backwards compatible.

`pkgs.formats` won't fit in a `mapAttrs (name: { type, generate }: foo)` anymore, or a `{ type, generate, generateText }:` for that matter.

I think I'm ok with that, but I would like @infinisil to review the idea.

File: nixos/modules/services/misc/gitea.nix
Comment (RaitoBezarius) [reply nixos/modules/services/misc/gitea.nix:234]: You can use mkDeprecatedOption or something like this IIRC

Comment (ThinkChaos) [reply nixos/modules/services/misc/gitea.nix]: Found `mkChangedOptionModule` and used that, hopefully that's what @SuperSandro2000 meant.

File: nixos/modules/services/misc/gitea.nix
Comment (infinisil) [inline nixos/modules/services/misc/gitea.nix:21]: I don't think `generateText` is a good idea for the formats, but it's also not needed, because you can just do
```suggestion
  configFile = pkgs.runCommand "app-extra.ini" {
    generated = format.generate "app.ini" cfg.settings;
    inherit extraConfig;
    passAsFile = [ "extraConfig" ];
  } ''
    cat "$generated" "$extraConfigPath"
  ''
```

(might need to add an extra newline in there)

File: pkgs/pkgs-lib/formats.nix
Comment (infinisil) [inline pkgs/pkgs-lib/formats.nix:152]: This is problematic for the same reasons discussed in https://github.com/NixOS/nixpkgs/issues/107491, which ultimately [led](https://github.com/NixOS/nixpkgs/pull/118925) to [`toINIWithGlobalSection`](https://github.com/NixOS/nixpkgs/pull/164088).

`pkgs.formats` should not add its own interface on top of that decision, but rather provide the same interface, which is apparently done in https://github.com/NixOS/nixpkgs/pull/201686.

**Review by infinisil — DISMISSED** [review]

File: pkgs/pkgs-lib/formats.nix
Comment (infinisil) [reply pkgs/pkgs-lib/formats.nix]: I don't think `generateText` is needed, see https://github.com/NixOS/nixpkgs/pull/206179#discussion_r1055695593

File: nixos/modules/services/misc/gitea.nix
Comment (roberth) [reply nixos/modules/services/misc/gitea.nix:21]: `generateText` _is_ relevant for the trusted evaluator kind of secrets deployment that nixops supports. That's not how it's used here though.

File: pkgs/pkgs-lib/formats.nix
Comment (ThinkChaos) [reply pkgs/pkgs-lib/formats.nix:152]: I hadn't thought of that limitation, thanks for the links.

I do think forcing the NixOS config to split the global and other sections would not be a nice user experience and want to avoid that.  
What do you think of having:
- `formats.ini`: `withGlobalSection` that splits `value` into `globalSection` and `sections` itself (opt-in of course)
- `formats.iniWithGlobalSection` that takes an attr with explicit `globalSection` and `section`, just like `generators.toINIWithGlobalSection`

`ini` would be the most common use case: you can't represent all potential configs, but anyways most software doesn't need to represent that edge case, so it's fine.  
`iniWithGlobalSection` would be used when you actually need to support that edge case.

If you don't think that's a good idea, then I can always implement the `settings` splitting in the `gitea` module (would also need to change the `settings` type to not rely on `(formats.ini {}).type`).
I think it'd be nice not to have to do that in every module, but up to you, I just want to merge my changes to Gitea 😄

File: nixos/modules/services/misc/gitea.nix
Comment (ThinkChaos) [reply nixos/modules/services/misc/gitea.nix:21]: IMO using `pkgs.runCommand` for this is a bit weird but it's not a strong opinion, so I'll update the MR.

Comment (infinisil) [reply nixos/modules/services/misc/gitea.nix:21]: We could also implement a `pkgs.concatFiles [ a b ]` which would just do `cat ${a} ${b}` and put the result in a new derivation. The fundamental idea of only reading the file at build time is key, since we don't actually need the string at eval-time.

Comment (infinisil) [reply nixos/modules/services/misc/gitea.nix:21]: @roberth I think the secret content should be concatenated using `cat` at runtime then, still not during evaluation

File: pkgs/pkgs-lib/formats.nix
Comment (infinisil) [reply pkgs/pkgs-lib/formats.nix:152]: I don't think it's a good idea to add custom interfaces on top. `pkgs.formats` should ideally just be an extended `lib.generators`. Either we add this splitting idea to `lib.generators` and `pkgs.formats` or neither. Now having the interface for `lib.generators.toINIWithGlobalSection` already decided, it makes most sense to just use the same interface for `pkgs.formats` too.

However I do agree that in a most cases, this splitting idea would enable a much nicer interface. I think it might be okay to have that _if_ we can make sure that we won't run into a situation where we can't represent a configuration. 

This essentially means: Check all the potential options of a whole bunch of programs using INI config files to see if you can find one that uses the same name for a global variable and a section. Maybe there's also an example where global variables are dynamic and can be freely picked, same for section names. You could also automate this by [looking at all INI files on GitHub](https://github.com/search?l=&p=3&q=language%3AINI&type=Code) to see if there's a clash. You might also want to check INI libraries in other programming languages to see how they handle global sections.

We need to be really sure because we'd essentially be redefining the semantics of global sections in the INI format. If we run into an example where NixOS can't represent some configuration in the future, we'd have to give pressure for the authors to change the config names to not clash.

Comment (ThinkChaos) [reply pkgs/pkgs-lib/formats.nix:152]: Put this on the back-burner for a while, and looking into it again now.

> Either we add this splitting idea to lib.generators and pkgs.formats or neither. Now having the interface for lib.generators.toINIWithGlobalSection already decided, it makes most sense to just use the same interface for pkgs.formats too.

I agree with this, but I'd prefer to make backwards compatible changes to the ini formats and generators so `formats.ini` can support splitting.  
Of course `ini` wouldn't support the conflicting global and section names edge case, but the key would be to make switching a module's format from `ini` to (a not yet created) `iniWithGlobalSection` transparent for the user's `settings`.

What I'm thinking is slightly changing  `toINIWithGlobalSection`'s semantics by changing the params  
from `{ globalSection, sections }` to `{ extraGlobals, values }`.  
`extraGlobals` would only accept global values (`attrsOf iniAtom`), as `globalSection` does, but it's intended use would mainly be for values whose name conflicts with a section.
`values` would be loosened compared to `sections` to accept both global values and sections (`either iniAtom (attrsOf iniAtom)`), and split into globals and sections.
Pseudo code to express the relationship between old and new params:  
- `globalSection = extractGlobals values // extraGlobals`
- `sections = extractSections values`

We'd have an assert to ensure `intersectAttrs (extractGlobals values) extraGlobals == { }`, so that nothing gets accidentally shadowed during the attrset merge.

This would make it possible to define `ini` as `iniWithGlobalSection` with `extraGlobals = { }`. Providing an upgrade path for any module that would someday need to support _the edge case_: switch the format from `ini` to `iniWithGlobalSection` and expose a second free-form attrset in the options that would feed be used as `extraGlobals`.  
Maybe something like so (this is just an example, and I'm not proposing a design for the module API):
```nix
{
  settings = { conflictingName = { ... }; };
  settingsExtraGlobals = { conflictingName = 0; };
}
```

How does that sound?

**Comment by ThinkChaos** [comment]
Rebased and removed the changes to formats and generators which I'll move to another PR.  
That should make this less intrusive and be good to merge as is AFAICT.

**Comment by RaitoBezarius** [comment]
> Rebased and removed the changes to formats and generators which I'll move to another PR. That should make this less intrusive and be good to merge as is AFAICT.

Thank you for your changes, I see some conflicts, could you also add a release notes mentioning that options are more "RFC042"-like I suppose?

**Comment by ThinkChaos** [comment]
Thanks for taking a look!
Force pushed to resolve conflicts and add release notes.

File: pkgs/pkgs-lib/formats.nix
Comment (infinisil) [reply pkgs/pkgs-lib/formats.nix]: Hmm it's an interesting idea, but I think we can make it work without pushing the edge case responsibility to all the modules. To do this we just need to detect whether there's a `global` keyword, and if so, request the user to distinguish between global sections and normal ones. More formally:
- Each definition of a `types.ini` option can either be of type `attrsOf (attrsOf iniAtom)` or `{ globalSection :: attrsOf iniAtom, sections :: attrsOf (attrsOf iniAtom) }`. In case of ambiguity, an appropriate error is thrown. Additionally if the type would be unambiguously the former one, but `globalSection` is a section name, an appropriate error is thrown. Internally the former gets coerced to the latter.

Here's some examples:
```nix
{
  settings = lib.mkMerge [
    # Valid, unambiguously the former type
    {
      foo.bar = 10;
      bar.baz = 20;
    }
    # Valid, unambiguously the latter type
    {
      globalSection.foo = 10;
      sections.bar.baz = 20;
    }
    # Invalid, ambiguous type
    {
      globalSection.foo = 20;
    }
    # Valid, only a global section, arguably the main drawback
    {
      globalSection.foo = 20;
      sections = {};
    }
  ];
}
```

File: nixos/doc/manual/release-notes/rl-2305.section.md
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2305.section.md:292]: > To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead.

File: nixos/modules/services/misc/gitea.nix
Comment (SuperSandro2000) [reply nixos/modules/services/misc/gitea.nix:315]: I think I meant omitting that the port is not shown in the web ui.

Comment (ThinkChaos) [reply nixos/modules/services/misc/gitea.nix:315]: It doesn't but it's already an improvement to the previous default that was the constant `http://localhost:3000/`.  
I think handling all cases is too complex to do here, was just trying to make it slightly more useful by default. I can revert it to the constant if you think it's better to not have any attempt to make it dynamic if it's not perfect.

File: nixos/doc/manual/release-notes/rl-2305.section.md
Comment (ThinkChaos) [reply nixos/doc/manual/release-notes/rl-2305.section.md:292]: Thanks, will do next time.

Comment (Ma27) [inline nixos/doc/manual/release-notes/rl-2305.section.md:292]: I wouldn't expect an average NixOS admin who's not involved with upstream development at all to know what RFC42 is about (despite the nice number :p)

Hence I'd prefer to provide a few more details, i.e. that a bunch of options don't exist anymore, but can be set with `services.gitea.settings` and that the keys match the upstream config params now.

File: nixos/modules/services/misc/gitea.nix
Comment (Ma27) [reply nixos/modules/services/misc/gitea.nix]: Considering that this won't change for existing installations anyways and the issue existed before, I'd resolve.

File: nixos/doc/manual/release-notes/rl-2305.section.md
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2305.section.md:322]:
```suggestion
- `gitea` module options have been changed to be RFC042 conforming (i.e. some options were moved to be located under `services.gitea.settings`)
```

otherwise people thing, that things would have been fully removed.
also please see the comment below the above headline.

File: nixos/modules/services/misc/gitea.nix
Comment (ThinkChaos) [reply nixos/modules/services/misc/gitea.nix:315]: Any further comment related to this @SuperSandro2000?  
This is the last unresolved thread.

**Comment by Ma27** [comment]
@ofborg test gitea

**Review by Ma27 — APPROVED** [review]
LGTM.

**Review by RaitoBezarius — APPROVED** [review]
LGTM.

**Comment by ThinkChaos** [comment]
I think this needs your approval @infinisil due to the `formats.nix` changes I reverted and will bring back in a follow-up PR.

**Review by FlorianFranzen — APPROVED** [review]
Looks great!

**Comment by ThinkChaos** [comment]
Thanks all for the reviews.

Am I supposed to do something to merge this?  
GitHub says "Only those with write access to this repository can merge pull requests."

File: nixos/modules/services/misc/gitea.nix
Comment (ThinkChaos) [reply nixos/modules/services/misc/gitea.nix]: Good catch indeed!

Pushed a fix, and also rebased.

**Comment by ThinkChaos** [comment]
Thanks, didn't mean to badger you, just wasn't sure what the practice was here!

**Review by RaitoBezarius — APPROVED** [review]

**Comment by roberth** [comment]
@ofborg test gitea

**Pull request merged**
